fontc/src/lib.rs

```diff
@@ -20,69 +23,86 @@ use std::{
-/// when the source is not a file on disk (WASM or library use).
-///
-/// Currently only supports Glyphs sources, but could be extended to e.g. UFOZ or other formats.
-#[derive(Debug, Clone, PartialEq)]
-pub enum InMemorySource {
-    Glyphs(String),
+#[cfg(feature = "cli")]
+pub use crate::timing::RealClock;
+
+pub enum Input {
+    DesignSpacePath(PathBuf),
+    GlyphsPath(PathBuf),
+    FontraPath(PathBuf),
+    GlyphsMemory(String),
 }
 
-/// Creates a [`Source`] from a source file loaded in memory
-fn create_in_memory_source(source: &InMemorySource) -> Result<Box<dyn Source>, Error> {
-    match source {
-        InMemorySource::Glyphs(source) => Ok(Box::new(GlyphsIrSource::new_from_memory(source)?)),
+impl Input {
+    pub fn new(path: &Path) -> Result<Self, Error> {
+        if !path.exists() {
+            return Err(Error::FileExpected(path.to_path_buf()));
+        }
+        let ext = path
+            .extension()
+            .and_then(OsStr::to_str)
+            .ok_or_else(|| Error::UnrecognizedSource(path.to_path_buf()))?;
+        match ext {
+            "designspace" => Ok(Input::DesignSpacePath(path.to_path_buf())),
+            "ufo" => Ok(Input::DesignSpacePath(path.to_path_buf())),
+            "glyphs" => Ok(Input::GlyphsPath(path.to_path_buf())),
+            "glyphspackage" => Ok(Input::GlyphsPath(path.to_path_buf())),
+            "fontra" => Ok(Input::FontraPath(path.to_path_buf())),
+            _ => Err(Error::UnrecognizedSource(path.to_path_buf())),
+        }
+    }
+
+    /// Creates a new input from a Glyphs source in memory
+    pub fn from_glyphs(source: String) -> Self {
+        Input::GlyphsMemory(source)
+    }
+
+    /// Creates the implementation of [`Source`] to feed to fontir.
+    fn create_source(&self) -> Result<Box<dyn Source>, Error> {
+        match self {
+            Input::DesignSpacePath(path) => Ok(Box::new(DesignSpaceIrSource::new(path)?)),
+            Input::GlyphsPath(path) => Ok(Box::new(GlyphsIrSource::new(path)?)),
+            Input::FontraPath(path) => Ok(Box::new(FontraIrSource::new(path)?)),
+            Input::GlyphsMemory(source) => Ok(Box::new(GlyphsIrSource::new_from_memory(source)?)),
+        }
     }
 }
 
-/// Run the compiler with the provided arguments
-pub fn run(args: Args, mut timer: JobTimer) -> Result<(), Error> {
-    let time = timer
-        .create_timer(AnyWorkId::InternalTiming("Init config"), 0)
-        .run();
-    let (ir_paths, be_paths) = init_paths(&args)?;
-    timer.add(time.complete());
+impl TryFrom<&Path> for Input {
+    type Error = Error;
 
-    let workload = Workload::new(args.clone(), timer)?;
+    fn try_from(path: &Path) -> Result<Self, Self::Error> {
+        Input::new(path)
+    }
+}
 
-    let fe_root = FeContext::new_root(args.flags(), ir_paths);
-    let be_root = BeContext::new_root(args.flags(), be_paths, &fe_root);
-    let mut timing = workload.exec(&fe_root, &be_root)?;
+#[cfg(feature = "cli")]
```

Comment:
no change required but generally i'm used to seeing any `#[attributes]` after any doc comments.